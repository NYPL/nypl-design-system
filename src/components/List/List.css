@import "../../styles/base/_03-mixins.css";

.list {
  /* browser automatically applies margin */
  margin: unset;

  &--no-list-styling {
    @add-mixin list-reset;
  }
}
