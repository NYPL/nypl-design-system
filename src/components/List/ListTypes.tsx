export enum ListTypes {
  Ordered = "ol",
  Unordered = "ul",
  Definition = "dl",
}
