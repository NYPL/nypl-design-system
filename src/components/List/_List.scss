@mixin list--bullet {
  color: var(--ui-gray-medium);
  content: "\2022"; /* \2022 is the CSS Code/unicode for a bullet */
  display: inline-block; /* Needed to add space between the bullet and the text */
  font-weight: bold;
  margin-left: -1em;
  width: 1em;
}

.list {
  margin: unset; // browser automatically applies margin

  &__definition-list {
    border-bottom: 1px solid var(--ui-gray-light);

    h2 {
      border-top: 3px solid var(--section-research-primary);
      margin: 0;
      padding: var(--space-xs) 0 0;
    }

    dl {
      display: grid;
      grid-template-columns: 100%;
      grid-template-rows: 1fr;
      margin: var(--space-xs) 0 0;

      dt {
        border-top: 1px solid var(--ui-gray-light);
        font-size: var(--font-size-0);
        font-weight: bold;
        padding-bottom: 0;
        padding-top: var(--space-s);

        @include breakpoint($breakpoint-medium) {
          padding-bottom: var(--space-s);
          padding-right: var(--space-m);
        }
      }

      dd {
        border-top: none;
        font-size: var(--font-size-0);
        margin: 0;
        padding-bottom: var(--space-s);

        @include breakpoint($breakpoint-medium) {
          border-top: 1px solid var(--ui-gray-light);
          padding-top: var(--space-s);
        }
      }

      @include breakpoint($breakpoint-medium) {
        grid-template-columns: max(225px) 1fr;
      }
    }
  }

  &--no-list-styling {
    @include list-reset;
  }
}

ul.list {
  list-style: none;

  li:before {
    @include list--bullet;
  }

  &--no-list-styling {
    li:before {
      content: unset; // Removes custom list bullet
    }
  }
}
