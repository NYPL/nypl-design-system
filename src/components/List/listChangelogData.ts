/** This data is used to populate the ComponentChangelogTable component.
 *
 * date: string (when adding new entry during development, set value as "Prerelease")
 * version: string (when adding new entry during development, set value as "Prerelease")
 * type: "Bug Fix" | "New Feature" | "Update";
 * affects: array["Accessibility" | "Documentation" | "Functionality" | "Styles"];
 * notes: array (will render as a bulleted list, add one array element for each list element)
 */
import { ChangelogData } from "../../utils/ComponentChangelogTable";

export const changelogData: ChangelogData[] = [
  {
    date: "Prerelease",
    version: "Prerelease",
    type: "Update",
    affects: ["Accessibility", "Documentation"],
    notes: [
      "Updated the `title` prop to allow JSX to render custom heading elements for accessible heading hierarchy.",
    ],
  },
];
