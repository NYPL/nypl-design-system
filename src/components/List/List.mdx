import { ArgTypes, Canvas, Description, Meta, Source } from "@storybook/blocks";

import * as ListStories from "./List.stories";
import Link from "../Link/Link";

<Meta of={ListStories} />

# List

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.7.0`    |
| Latest            | `1.7.2`    |

## Table of Contents

- {<Link href="#overview" target="_self">Overview</Link>}
- {<Link href="#component-props" target="_self">Component Props</Link>}
- {<Link href="#accessibility" target="_self">Accessibility</Link>}
- {<Link href="#description-list" target="_self">Description List</Link>}
- {<Link href="#description-list-of-links" target="_self">Description List of Links</Link>}
- {<Link href="#lists-with-data-props" target="_self">Lists with Data Props</Link>}

## Overview

<Description of={ListStories} />

<Canvas of={ListStories.WithControls} />

<ArgTypes of={ListStories.WithControls} />

## Accessibility

Internally, the `List` component implements native HTML markup for unordered,
ordered, and description lists. This includes `ul`, `ol`, and `dl` elements.
When data is passed into the `listItems` data prop, the appropriate children
HTML elements are used. This includes `li` for unorder and ordered lists, and
`dt` and `dd` elements for description lists.

The description type for the `List` component is wrapped in a `section` element
with an `h2` header. Keep that in mind when ordering headings on the page.

Resources:

- [W3C WAI Content Structure](https://www.w3.org/WAI/tutorials/page-structure/content/)
- [Deque University Appropriate Semantic Markup for Lists](https://dequeuniversity.com/tips/list-markup)

## Description List

_Note: This element is officially called the "Description List" element in HTML5.
Before HTML5, it was called the "Definition List" element and some online resources
may still use this name. In the Reservoir Design System, we will use the
"Description List" name._

- [W3C Using description lists](https://www.w3.org/WAI/WCAG21/Techniques/html/H40.html)
- [MDN dl: The Description List element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl)

To render a description list, pass in `"dl"` to the `type` prop. The optional
`title` prop will now render above the description list element. This type of
list renders `dt` and `dd` elements.

<Source
  code={`
<List type="dl">
  <dt>Term</dt>
  <dd>Description</dd>
</List>
`}
  language="jsx"
/>

<Canvas of={ListStories.DescriptionList} />

## Description List of Links

An example with HTML elements inside of the `dd` elements.

<Canvas of={ListStories.DescriptionListWithLinks} />

## Lists with Data Props

If `li` or `dt`/`dd` elements cannot be passed as children to the `List` element,
an array with data can be passed into the `listItems` prop. Note: either
children elements or the `listItems` prop should be passed but not both.

### Unordered and Ordered

For lists of type `"ul"` (unorderd) or `"ol"` (ordered), `li` list item elements
can be passed as children. If that's not possible but the data to render is
stored as an array of strings, then that data array can be passed into `listItems`.

<Source
  code={`
// With \`li\` elements
<List type={ul}>
  <li>Mahi-mahi</li>
  <li>Golden trout</li>
  <li>Rainbowfish</li>
  <li>Suckerfish</li>
</List>
`}
  language="jsx"
/>

<Source
  code={`
// With \`itemList\` data prop
const fishArray = ["Mahi-mahi", "Golden trout", "Rainbowfish", "Suckerfish"];
<List type={ul} listItems={fishArray} />
`}
  language="jsx"
/>

### Description

For lists of type `"dl"` (description), `dt` and `dd` elements can be
passed as children. If that's not possible but the data to render is stored as
an array of objects, then that data array can be passed into `listItems`. The
object must have two keys, `term` and `description`.

<Source
  code={`
// With \`dt\`/\`dd\` elements
<List type="dl" title="Animal Crossing Fish">
  <dt>Mahi-mahi</dt>
  <dd>The mahi-mahi is an ocean fish known...</dd>
  <dt>Golden trout</dt>
  <dd>The golden trout is a beautifully colored fish...</dd>
  <dt>Rainbowfish</dt>
  <dd>The rainbowfish is a tropical fish known...</dd>
  <dt>Suckerfish</dt>
  <dd>The suckerfish is a curious fish that...</dd>
</List>
`}
  language="jsx"
/>

<Source
  code={`
// With \`itemList\` data prop
const fishDescriptions = [
  {
    term: "Mahi-mahi",
    description: "The mahi-mahi is an ocean fish known..."
  },
  {
    term: "Golden trout",
    description: "The golden trout is a beautifully colored fish..."
  },
  {
    term: "Rainbowfish",
    description: "The rainbowfish is a tropical fish known..."
  },
  {
    term: "Suckerfish"
    description: "he suckerfish is a curious fish that..."
  }
];
// ...
<List listItems={fishDescriptions} title="Animal Crossing Fish" type="dl" />
`}
  language="jsx"
/>

<Canvas of={ListStories.ListWithDataProps} />
