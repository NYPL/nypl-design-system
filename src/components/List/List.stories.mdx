import {
  ArgsTable,
  Canvas,
  Description,
  Meta,
  Story,
} from "@storybook/addon-docs";
import { withDesign } from "storybook-addon-designs";

import Heading from "../Heading/Heading";
import { HeadingLevels } from "../Heading/HeadingTypes";
import Image from "../Image/Image";
import Link from "../Link/Link";
import { LinkTypes } from "../Link/LinkTypes";
import List from "./List";
import { ListTypes } from "./ListTypes";
import { getCategory } from "../../utils/componentCategories";
import { getStorybookEnumValues } from "../../utils/utils";

export const enumValues = getStorybookEnumValues(ListTypes, "ListTypes");

<Meta
  title={getCategory("List")}
  component={List}
  decorators={[withDesign]}
  parameters={{
    design: {
      type: "figma",
      url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Master?node-id=16115%3A304",
    },
    jest: ["List.test.tsx"],
  }}
  argTypes={{
    additionalStyles: { control: false },
    children: { table: { disable: true } },
    className: { control: false },
    id: { control: false },
    inline: { table: { defaultValue: { summary: false } } },
    listItems: { control: false },
    noStyling: { table: { defaultValue: { summary: false } } },
    type: {
      control: { type: "select" },
      options: enumValues.options,
      table: { defaultValue: { summary: "ListTypes.Unordered" } },
    },
  }}
/>

# List

| Component Version | DS Version |
| ----------------- | ---------- |
| Added             | `0.7.0`    |
| Latest            | `0.26.0`   |

## Table of Contents

- [Overview](#overview)
- [Component Props](#component-props)
- [Accessibility](#accessibility)
- [Description List](#description-list)
- [Description List of Links](#description-list-of-links)
- [Lists with Data Props](#lists-with-data-props)

## Overview

<Description of={List} />

export const animalCrossingDescriptions = [
  {
    term: "Mahi-mahi",
    description:
      'The mahi-mahi is an ocean fish known for its wide, somewhat-cute face. It can reach over six feet long. It is known by different names including "dolphinfish," even though it has no relation to dolphins. They live in only in warm, tropical waters...which perhaps explains the relaxed, happy look on their faces.',
  },
  {
    term: "Golden trout",
    description: (
      <p>
        The golden trout is a <b>beautifully</b> colored fish that can only live
        in very clean waters. They are difficult to come across since they are
        found only in high mountain streams. As a side note, I find it much
        easier to appreciate fish that aren't such prima donnas about
        everything.
      </p>
    ),
  },
  {
    term: "Rainbowfish",
    description:
      'The rainbowfish is a tropical fish known for its metallic colors and beautiful fins. There are over 50 different species, each unique and pleasing color. I must say, it does make me wish for feathers of a more exciting hue than "underbaked brownie."',
  },
  {
    term: "Suckerfish",
    description: (
      <p>
        The{" "}
        <Link href="https://animalcrossing.fandom.com/wiki/Suckerfish">
          suckerfish
        </Link>{" "}
        is a curious fish that likes to attach itself with its sucker mouth to
        larger marine animals. The benefit to the suckerfish is that it gets to
        eat smaller parasites and dead skin off the host's body. Amusingly, some
        people have used suckerfish on cords to catch large turtles with the
        fish's own suction! I imagine this practice is the cause of some awkward
        conversations between fish and turtle, eh wot?
      </p>
    ),
  },
];
export const itemGroups = [
  "Art",
  "Bushes",
  "Clothing",
  "DIY Recipes",
  "Flowers",
  "Fossils",
  "Fruits",
  "Furniture",
  "Materials",
  "NPC",
  "Songs",
  "Tools",
  "Villagers",
];
export const descriptions = [
  {
    term: "Balrogs",
    description:
      'Demonic creatures of fire and shadow, Balrogs were fallen Maiar, loyal to the first Dark Lord, Morgoth. They participated in the wars of the First Age of Middle-earth but were mostly destroyed during the War of Wrath which ended the Age. By the Third Age, the only remaining Balrog was "Durin\'s Bane," the Balrog of Moria, killed by Gandalf.',
  },
  {
    term: "Dwarves",
    description:
      "The race of Dwarves preferred to live in mountains and caves, settling in places such as Erebor (the Lonely Mountain), the Iron Hills, the Blue Mountains, and Moria (Khazad-dûm) in the Misty Mountains. Aulë the Smith created Dwarves; he also invented the Dwarven language, known as Khuzdul. Dwarves mined and worked precious metals throughout the mountains of Middle-earth. The seven different groups of Dwarf-folk originated in the locations where the Seven Fathers of the Dwarves first awoke before the First Age.",
  },
  {
    term: "Elves",
    description:
      'The Elves, or Firstborn, were the first of Eru\'s Children to awaken. Born under the stars before the ascension of the Moon and the Sun, they retain a special love for light and an inner spirit endowed with unique gifts. They call themselves the Quendi, or "Speakers", for they were the first to utter words; and, even now, no race understands language and song like the Firstborn. Fair and fine featured, brilliant and proud, immortal and strong, tall and agile, they are the most blessed of the Free Peoples. They can see as well under moon or starlight as a man at the height of day. They cannot become sick or scarred, but if an Elf should die, from violence or losing the will to live from grief, their spirit goes to the halls of Mandos, and as they are bound to Arda and cannot leave until the world is broken and remade. Elven skill and agility is legendary: for instance, walking atop freshly fallen snow without leaving a trace of their passing. On a clear day they can see ten miles with perfect clarity and detail up to 100 miles. These gifts come at great cost, though: they are strongly bound to Fate (see Mandos) and hated by Morgoth. No other race has been blessed and cursed more than the Quendi.',
  },
  {
    term: "Ents",
    description:
      "Ents were an ancient race of tree-like creatures, having become like the trees that they shepherd. They were created by Yavanna and given life by Ilúvatar. By the Third Age, they were a dwindling race, having long ago lost their mates, the Entwives.",
  },
  {
    term: "Hobbits",
    description:
      'Hobbits are a race of Middle-earth, also known as "halflings" on account of their short stature, roughly half the size of men. They are characterized by curly hair on their heads and leathery feet that have furry insteps, for which they did not wear shoes. Many hobbits live in the Shire as well as Bree, and they once lived in the vales of the Anduin. They are fond of an unadventurous life of farming, eating, and socializing. There were three types of Hobbits: The Harfoots were the most numerous. The Stoors had an affinity for water, boats and swimming; the Fallohides were an adventurous people. The origin of hobbits is unclear, but of all the races they have the closest affinity to men, and in the Prologue to The Lord of the Rings Tolkien calls them relatives of men.',
  },
];

<Canvas withToolbar>
  <Story
    name="List with Controls"
    args={{
      additionalStyles: undefined,
      className: undefined,
      id: "nypl-list",
      inline: false,
      listItems: undefined,
      noStyling: false,
      title: "Middle-Earth Peoples",
      type: "ListTypes.Unordered",
    }}
  >
    {(args) => (
      <List {...args} type={enumValues.getValue(args.type)}>
        {args.type !== "ListTypes.Description"
          ? itemGroups.map((item, i) => <li key={i}>{item}</li>)
          : descriptions.map((item, i) => [
              <dt key={`dt_${i}`}>{item.term}</dt>,
              <dd key={`dd_${i}`}>{item.description}</dd>,
            ])}
      </List>
    )}
  </Story>
</Canvas>

<ArgsTable story="List with Controls" />

## Accessibility

Internally, the `List` component implements native HTML markup for unordered,
ordered, and description lists. This includes `ul`, `ol`, and `dl` elements.
When data is passed into the `listItems` data prop, the appropriate children
HTML elements are used. This includes `li` for unorder and ordered lists, and
`dt` and `dd` elements for description lists.

The description type for the `List` component is wrapped in a `section` element
with an `h2` header. Keep that in mind when ordering headings on the page.

Resources:

- [W3C WAI Content Structure](https://www.w3.org/WAI/tutorials/page-structure/content/)
- [Deque University Appropriate Semantic Markup for Lists](https://dequeuniversity.com/tips/list-markup)

## Description List

_Note: This element is officially called the "Description List" element in HTML5.
Before HTML5, it was called the "Definition List" element and some online resources
may still use this name. In the Reservoir Design System, we will use the
"Description List" name._

- [W3C Using description lists](https://www.w3.org/WAI/WCAG21/Techniques/html/H40.html)
- [MDN dl: The Description List element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl)

To render a description list, pass in `ListTypes.Description` to the `type` prop.
The optional `title` prop will now render above the description list element.
This type of list renders `dt` and `dd` elements.

```jsx
<List type={ListTypes.Description}>
  <dt>Term</dt>
  <dd>Description</dd>
</List>
```

<Canvas>
  <Story
    name="Description List"
    args={{
      id: "nypl-list2",
      noStyling: false,
      title: "Middle-Earth Peoples",
    }}
    argTypes={{
      inline: { control: false },
      noStyling: { control: false },
      type: { control: false },
    }}
    parameters={{
      design: {
        type: "figma",
        url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?node-id=10734%3A5178",
      },
    }}
  >
    {(args) => (
      <List {...args} type={ListTypes.Description}>
        {descriptions.map((item, i) => [
          <dt key={`dt_${i}`}>{item.term}</dt>,
          <dd key={`dd_${i}`}>{item.description}</dd>,
        ])}
      </List>
    )}
  </Story>
</Canvas>

## Description List of Links

An example with HTML elements inside of the `dd` elements.

<Canvas>
  <Story
    name="Description List of Links"
    args={{
      noStyling: false,
      title: "Details",
    }}
    argTypes={{
      inline: { control: false },
      noStyling: { control: false },
      type: { table: { disable: true } },
    }}
    parameters={{
      design: {
        type: "figma",
        url: "https://www.figma.com/file/qShodlfNCJHb8n03IFyApM/Main?node-id=10734%3A5178",
      },
    }}
  >
    {(args) => (
      <List {...args} type={ListTypes.Description}>
        <dt>Authors</dt>
        <dd>
          <a href="#">Chirwa, Ephraim Wadonda, author</a>
          <br />
          <a href="#">Dorward, Andrew, author</a>
        </dd>
        <dt>Subjects</dt>
        <dd>
          <a href="#">Ackerbau</a>
          <br />
          <a href="#">Ackerbau.</a>
          <br />
          <a href="#">Africa.</a>
          <br />
          <a href="#">Afrika.</a>
          <br />
          <a href="#">Agrarsubvention</a>
          <br />
          <a href="#">Agrarsubvention.</a>
          <br />
          <a href="#">Agricultura -- Subvencions -- Malawi.</a>
          <br />
          <a href="#">Agricultural economics New.</a>
          <br />
          <a href="#">Agricultural subsidies</a>
          <br />
          <a href="#">Agricultural subsidies -- Developing countries.</a>
          <br />
          <a href="#">Agricultural subsidies -- Malawi.</a>
          <br />
          <a href="#">Agricultural subsidies.</a>
          <br />
          <a href="#">BUSINESS &amp; ECONOMICS -- Industries</a>
          <br />
        </dd>
      </List>
    )}
  </Story>
</Canvas>

## Lists with Data Props

If `li` or `dt`/`dd` elements cannot be passed as children to the `List` element,
an array with data can be passed into the `listItems` prop. Note: either
children elements or the `listItems` prop should be passed but not both.

### Unordered and Ordered

For lists of type `ListTypes.Unordered` (ul) or `ListTypes.Ordered` (ol), `li`
list item elements can be passed as children. If that's not possible but the
data to render is stored as an array of strings, then that data array can be
passed into `listItems`.

```jsx
<!-- With `li` elements -->
<List type={ListTypes.Unordered}>
  <li>Mahi-mahi</li>
  <li>Golden trout</li>
  <li>Rainbowfish</li>
  <li>Suckerfish</li>
</List>

<!-- With `itemList` data prop -->
const fishArray = ["Mahi-mahi", "Golden trout", "Rainbowfish", "Suckerfish"];
<List type={ListTypes.Unordered} listItems={fishArray} />
```

### Description

For lists of type `ListTypes.Description` (dl), `dt` and `dd` elements can be
passed as children. If that's not possible but the data to render is stored as
an array of objects, then that data array can be passed into `listItems`. The
object must have two keys, `term` and `description`.

```jsx
// With `dt`/`dd` elements
<List type={ListTypes.Description} title="Animal Crossing Fish">
  <dt>Mahi-mahi</dt>
  <dd>The mahi-mahi is an ocean fish known...</dd>
  <dt>Golden trout</dt>
  <dd>The golden trout is a beautifully colored fish...</dd>
  <dt>Rainbowfish</dt>
  <dd>The rainbowfish is a tropical fish known...</dd>
  <dt>Suckerfish</dt>
  <dd>The suckerfish is a curious fish that...</dd>
</List>

// With `itemList` data prop
const fishDescriptions = [
  {
    term: "Mahi-mahi",
    description: "The mahi-mahi is an ocean fish known..."
  },
  {
    term: "Golden trout",
    description: "The golden trout is a beautifully colored fish..."
  },
  {
    term: "Rainbowfish",
    description: "The rainbowfish is a tropical fish known..."
  },
  {
    term: "Suckerfish"
    description: "he suckerfish is a curious fish that..."
  }
];
<List
  listItems={fishDescriptions}
  title="Animal Crossing Fish"
  type={ListTypes.Description}
/>
```

<Canvas>
  <Story
    name="List with Data Props"
    args={{
      listItems: animalCrossingDescriptions,
      noStyling: false,
      title: "Animal Crossing Fish",
    }}
    argTypes={{
      inline: { control: false },
      noStyling: { control: false },
      type: { control: false },
    }}
  >
    {(args) => <List {...args} type={ListTypes.Description} />}
  </Story>
</Canvas>
