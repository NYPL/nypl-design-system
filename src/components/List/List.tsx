import * as React from "react";
import bem from "../../utils/bem";
import { ListTypes } from "./ListTypes";
import Heading from "../Heading/Heading";

export interface ListProps {
  /** BlockName for use with BEM. See how to work with blockNames and BEM here: http://getbem.com/introduction/ */
  blockName?: string;
  /** ClassName you can add in addition to 'list' */
  className?: string;
  /** ID that other components can cross reference for accessibility purposes */
  id?: string;
  /** Modifiers array for use with BEM. See how to work with modifiers and BEM here: http://getbem.com/introduction/ */
  modifiers?: any[];
  /** Ordered, Unordered, or Definition */
  type: ListTypes;
  /** An optional title that will appear over the list (only applies to Definition Lists) */
  title?: string;
}

export default function List(props: React.PropsWithChildren<ListProps>) {
  const {
    blockName,
    children,
    className,
    id,
    modifiers = [],
    type,
    title,
  } = props;

  const baseClass = "list";

  let listTag;
  let errorText = "";

  switch (type) {
    case ListTypes.Ordered:
      errorText = "Direct children of `List` (ordered) should be `<li>`s";
      React.Children.map(children, function (child: React.ReactElement) {
        if (child.type !== "li") {
          throw new Error(errorText);
        }
      });
      listTag = (
        <ol
          id={id}
          className={bem(baseClass, modifiers, blockName, [className])}
        >
          {children}
        </ol>
      );
      break;
    case ListTypes.Unordered:
      errorText = "Direct children of `List` (unordered) should be `<li>`s";
      React.Children.map(children, function (child: React.ReactElement) {
        if (child.type !== "li") {
          throw new Error(errorText);
        }
      });
      listTag = (
        <ul
          id={id}
          className={bem(baseClass, modifiers, blockName, [className])}
        >
          {children}
        </ul>
      );
      break;
    case ListTypes.Definition:
      errorText =
        "Direct children of `List` (definition) should be `<dt>`s or `<dd>`s";
      React.Children.map(children, function (child: React.ReactElement) {
        if (
          child.type !== "dt" &&
          child.type !== "dd" &&
          child.type !== React.Fragment
        ) {
          throw new Error(errorText);
        }
      });
      listTag = (
        <section
          id={id}
          className={bem("definition-list", modifiers, baseClass)}
        >
          {title && (
            <Heading id={`${id}-heading`} level={2}>
              {title}
            </Heading>
          )}
          <dl>{children}</dl>
        </section>
      );
      break;
  }

  return listTag;
}
