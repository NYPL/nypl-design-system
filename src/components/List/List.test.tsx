import { expect } from "chai";

import * as Enzyme from "enzyme";
import * as React from "react";

import List from "./List";
import { ListTypes } from "./ListTypes";

describe("List", () => {
  let wrapper: Enzyme.ReactWrapper<any, any>;

  it("Returns unordered list", () => {
    wrapper = Enzyme.mount(
      <List type={ListTypes.Unordered}>
        <li>hello</li>
      </List>
    );
    expect(wrapper.find("ul").exists()).to.equal(true);
    expect(wrapper.find("ul").hasClass("list")).to.equal(true);
    expect(wrapper.find("li").exists()).to.equal(true);
  });

  it("Returns ordered list", () => {
    wrapper = Enzyme.mount(
      <List type={ListTypes.Ordered}>
        <li>hello</li>
      </List>
    );
    expect(wrapper.find("ol").exists()).to.equal(true);
    expect(wrapper.find("ol").hasClass("list")).to.equal(true);
    expect(wrapper.find("li").exists()).to.equal(true);
  });

  it("Returns definition list", () => {
    wrapper = Enzyme.mount(
      <List type={ListTypes.Definition}>
        <dt>Dictionary Term</dt>
        <dd>Dictionary Definition</dd>
      </List>
    );
    expect(wrapper.find("dl").exists()).to.equal(true);
    expect(wrapper.find("dt").exists()).to.equal(true);
    expect(wrapper.find("dd").exists()).to.equal(true);
  });

  it("Throws an error when you pass an ordered or unordered list children that aren't <li>s", () => {
    expect(
      () =>
        (wrapper = Enzyme.mount(
          <List type={ListTypes.Ordered}>
            <span>hello</span>
            <span>hello</span>
            <span>hello</span>
          </List>
        ))
    ).to.throw("Direct children of `List` (ordered) should be `<li>`s");
  });

  it("Throws an error when you pass a definition list children that aren't `<dt>`s or `<dd>`s", () => {
    expect(
      () =>
        (wrapper = Enzyme.mount(
          <List type={ListTypes.Definition}>
            <span>hello</span>
            <span>hello</span>
            <span>hello</span>
          </List>
        ))
    ).to.throw(
      "Direct children of `List` (definition) should be `<dt>`s or `<dd>`s"
    );
  });
});
