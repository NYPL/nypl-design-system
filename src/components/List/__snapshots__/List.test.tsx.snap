// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Renders the UI snapshot correctly 1`] = `
<ul
  className="css-1xdhyk6"
  id="unordered"
>
  <li>
    Mahi-mahi
  </li>
  <li>
    Golden trout
  </li>
  <li>
    Rainbowfish
  </li>
  <li>
    Suckerfish
  </li>
</ul>
`;

exports[`List Renders the UI snapshot correctly 2`] = `
<ul
  className="css-1xdhyk6"
  id="ordered"
>
  <li>
    Mahi-mahi
  </li>
  <li>
    Golden trout
  </li>
  <li>
    Rainbowfish
  </li>
  <li>
    Suckerfish
  </li>
</ul>
`;

exports[`List Renders the UI snapshot correctly 3`] = `
<ol
  className="css-1xdhyk6"
  id="unordered-no-styling"
>
  <li>
    Mahi-mahi
  </li>
  <li>
    Golden trout
  </li>
  <li>
    Rainbowfish
  </li>
  <li>
    Suckerfish
  </li>
</ol>
`;

exports[`List Renders the UI snapshot correctly 4`] = `
<ol
  className="css-1xdhyk6"
  id="ordered-no-styling"
>
  <li>
    Mahi-mahi
  </li>
  <li>
    Golden trout
  </li>
  <li>
    Rainbowfish
  </li>
  <li>
    Suckerfish
  </li>
</ol>
`;

exports[`List Renders the UI snapshot correctly 5`] = `
<section
  className="css-1xdhyk6"
  id="description"
>
  <h2
    className="chakra-heading css-878tmy"
    id="description-heading"
  >
    Animal Crossing Fish
  </h2>
  <dl>
    <dt>
      Mahi-mahi
    </dt>
    <dd>
      <p>
        The mahi-mahi is an ocean fish known...
      </p>
    </dd>
    <dt>
      Golden trout
    </dt>
    <dd>
      The golden trout is a beautifully colored fish...
    </dd>
  </dl>
</section>
`;

exports[`List Renders the UI snapshot correctly 6`] = `
<ul
  className="css-kle7zy"
  id="chakra"
>
  <li>
    Mahi-mahi
  </li>
  <li>
    Golden trout
  </li>
  <li>
    Rainbowfish
  </li>
  <li>
    Suckerfish
  </li>
</ul>
`;

exports[`List Renders the UI snapshot correctly 7`] = `
<ul
  className="css-1xdhyk6"
  data-testid="other"
  id="other"
>
  <li>
    Mahi-mahi
  </li>
  <li>
    Golden trout
  </li>
  <li>
    Rainbowfish
  </li>
  <li>
    Suckerfish
  </li>
</ul>
`;

exports[`List Renders the UI snapshot correctly 8`] = `
<section
  className="css-kle7zy"
  id="chakra"
>
  <h2
    className="chakra-heading css-878tmy"
    id="chakra-heading"
  >
    Animal Crossing Fish
  </h2>
  <dl>
    <dt>
      Mahi-mahi
    </dt>
    <dd>
      <p>
        The mahi-mahi is an ocean fish known...
      </p>
    </dd>
    <dt>
      Golden trout
    </dt>
    <dd>
      The golden trout is a beautifully colored fish...
    </dd>
  </dl>
</section>
`;

exports[`List Renders the UI snapshot correctly 9`] = `
<section
  className="css-1xdhyk6"
  data-testid="other"
  id="other"
>
  <h2
    className="chakra-heading css-878tmy"
    id="other-heading"
  >
    Animal Crossing Fish
  </h2>
  <dl>
    <dt>
      Mahi-mahi
    </dt>
    <dd>
      <p>
        The mahi-mahi is an ocean fish known...
      </p>
    </dd>
    <dt>
      Golden trout
    </dt>
    <dd>
      The golden trout is a beautifully colored fish...
    </dd>
  </dl>
</section>
`;
